contracts/misc/src/tag/Tag.sol

```diff
@@ -0,0 +1,748 @@
+
+        emit CrosschainSend(destChainId, msg.sender, to, tokenId);
+    }
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                      PORTAL FUNCTIONS                      */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    function processCrosschainMint(address from, address to, uint256 amount, address refundTo, bool payGas)
+        public
+        xrecv
+    {
+        if (xmsg.sender != address(this)) revert Unauthorized();
+
+        // Determine how many NFTs can be minted
+        (uint256 validAmount, bool validMint) = _calculateValidMintAmount(to, amount, false);
+
+        // Only mint the valid amount
+        if (validMint) _processMint(from, to, validAmount, false);
+
+        if (!validMint) {
+            // Refund if no NFTs can be minted
+            uint256 refundAmount = amount * price;
+            _refundNative(refundTo, 0, refundAmount);
+            emit CrosschainMintFailRefund(refundTo, refundAmount);
+        } else if (validAmount < amount) {
+            // Refund for NFTs that couldn't be minted
+            uint256 refundAmount = (amount - validAmount) * price;
+            _refundNative(refundTo, 0, refundAmount);
+            emit CrosschainMintPartialFailRefund(refundTo, refundAmount);
+        }
+
+        // Pay gas if requested
+        if (payGas) _sendGas(to);
+    }
+
+    function processCrosschainMintWhitelist(
+        address from,
+        address to,
+        uint256 amount,
+        bytes32[] calldata proof,
+        address refundTo,
+        bool payGas
+    ) public xrecv {
+        if (xmsg.sender != address(this)) revert Unauthorized();
+
+        bool whitelistAvailable = _validateWhitelist(from, proof, true);
+        if (!whitelistAvailable) {
+            // If whitelist claim isn't available, process as normal mint
+            processCrosschainMint(from, to, amount - 1, refundTo, payGas);
+            return;
+        }
+
+        // Determine how many NFTs can be minted
+        (uint256 validAmount, bool validMint) = _calculateValidMintAmount(to, amount, true);
+
+        // Only mint the valid amount
+        if (validMint) _processMint(from, to, validAmount, true);
+
+        if (!validMint) {
+            // Refund if no paid for NFTs can be minted
+            if (amount > 1) {
+                uint256 refundAmount = (amount - 1) * price;
+                _refundNative(refundTo, 0, refundAmount);
+                emit CrosschainMintFailRefund(refundTo, refundAmount);
+            }
+        } else if (validAmount < amount && validAmount > 1) {
+            // Refund for NFTs that couldn't be minted
+            uint256 refundAmount = (amount - validAmount - 1) * price;
+            _refundNative(refundTo, 0, refundAmount);
+            emit CrosschainMintPartialFailRefund(refundTo, refundAmount);
+        }
+
+        // Pay gas if requested
+        if (payGas) _sendGas(to);
```

Comment:
Couple things on _sendGas?

- how do we make sure we have enough to service requests? prefunding, monitoring, and funding if needed?
- if transfer in _sendGas fails, then the call reverts. NFT is not minted and user is not refunded. Is that what we want?